remotes/fixup.go

```diff
@@ -212,21 +247,109 @@ func fixupImage(ctx context.Context, baseImage bundle.BaseImage, cfg fixupConfig
 	return baseImage, nil
 }
 
-// fixupPlatform resolve a single image manifest out of a manifest list if a platform filter has been specified
-// it modifies the baseImage and fixupInfo accordingly
-func fixupPlatform(ctx context.Context, cfg fixupConfig, baseImage *bundle.BaseImage, fixupInfo *imageFixupInfo, sourceFetcher remotes.Fetcher) error {
-	if cfg.platform == "" ||
+func fixupPlatforms(ctx context.Context, cfg fixupConfig, baseImage *bundle.BaseImage, fixupInfo *imageFixupInfo, sourceFetcher sourceFetcherAdder, filter []platforms.Matcher) error {
+	if len(filter) == 0 ||
 		(fixupInfo.resolvedDescriptor.MediaType != ocischemav1.MediaTypeImageIndex && fixupInfo.resolvedDescriptor.MediaType != images.MediaTypeDockerSchema2ManifestList) {
 		// no platform filter if platform is empty, or if the descriptor is not an OCI Index / Docker Manifest list
 		return nil
 	}
+	if len(filter) == 1 {
+		return fixupSinglePlatform(ctx, cfg, baseImage, fixupInfo, sourceFetcher, filter[0])
+	}
 
-	plat, err := platforms.Parse(cfg.platform)
+	reader, err := sourceFetcher.Fetch(ctx, fixupInfo.resolvedDescriptor)
 	if err != nil {
 		return err
 	}
-	matcher := platforms.NewMatcher(plat)
+	defer reader.Close()
 
+	manifestBytes, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return err
+	}
+	var manifestList typelessManifestList
+	if err := json.Unmarshal(manifestBytes, &manifestList); err != nil {
+		return err
+	}
+	valid := 0
+	for _, d := range manifestList.Manifests {
+		if matchAny(d.Platform, filter) {
+			manifestList.Manifests[valid] = d
+			valid++
+		}
+	}
+	manifestList.Manifests = manifestList.Manifests[:valid]
```

Comment:
These half dozen lines seem a complicated way to write:
```go
var manifests []themanifesttype // make([]..., 0, len(manifestList.Manifests)) is you really want to optimise the array alloc.
for _, d := range manifestList.Manifests {
		if matchAny(d.Platform, filter) {
			manifests = append(manifests, d)
		}
	}
}
manifestList.Manifests = new
```
(not adjusted for `platforms.Any` suggestion above)

Reply:
this is the pattern I use for in-place filtering, but ok :)